### PR TITLE
WT-11970 Do not raise an error if background compaction is already enabled/disabled

### PR DESCRIPTION
**Pull request by Sean04**

### CODE REVIEW

File: src/conn/conn_compact.c [hunk 1]
```diff
@@ -684,15 +684,18 @@ __wt_background_compact_signal(WT_SESSION_IMPL *session, const char *config)
...
-        /*
-         * This is an error as we are already in the same state and reconfiguration is not allowed.
-         */
+    toggle = cval.val;
+
+    /* Strip the unused fields from the configuration to check if the configuration has changed. */
+    WT_ERR(__wt_config_merge(session, cfg, "background=,exclude=", &stripped_config));
+
+    /* The background compact configuration cannot be changed while it's already running. */
+    if (toggle && running && strcmp(stripped_config, conn->background_compact.config) != 0)
```
Comment (etienneptl) [inline src/conn/conn_compact.c:693]: As discussed, please early exit if we are staying in the same state.

File: test/suite/test_compact06.py [hunk 2]
```diff
@@ -52,15 +52,11 @@ def test_background_compact_api(self):
                 self.session.compact(None, f'background=false,{item}'),
                 '/configuration cannot be set when disabling the background compaction server/')
 
-        #   3. We cannot disable the background server when it is already disabled.
-        self.assertRaisesWithMessage(wiredtiger.WiredTigerError, lambda:
-            self.session.compact(None, 'background=false'), '/Background compaction is already disabled/')
-
-        #   4. We cannot enable the background server with an invalid URIs to be excluded.
+        #   3. We cannot enable the background server with an invalid URIs to be excluded.
```
Comment (etienneptl) [inline test/suite/test_compact06.py:55]: Sorry, it's not you but there is a typo here URI*s*:
```suggestion
        #   3. We cannot exclue invalid URIs when enabling background compaction.
```

File: src/conn/conn_compact.c (hunk 1 above)
Comment (etienneptl) [reply src/conn/conn_compact.c:693]: At the end of the function you are changing, there are those lines:
```
    if (ret == 0)
        __wt_cond_signal(session, conn->background_compact.cond);
```
I don't think we should execute them either if we are in the same state.

Comment (Sean04) [reply src/conn/conn_compact.c:693]: That's true, I don't think that's problematic though. Should we change it regardless?
```suggestion
    __wt_spin_unlock(session, &conn->background_compact.lock);
    return (ret);
```

Comment (etienneptl) [reply src/conn/conn_compact.c:693]: I would like to keep the locks in one place, i.e after the `err` label.
@marcbutler, what do you think we should do?

Comment (etienneptl) [reply src/conn/conn_compact.c:693]: Moving it outside the `err` label should be good enough to tackle this.

**Review by etienneptl — APPROVED** [review]
LGTM!

File: src/conn/conn_compact.c (hunk 1 above)
Comment (marcbutler) [reply src/conn/conn_compact.c:693]: I agree minimizing the paths involved in locking makes it much easier to verify.

File: src/conn/conn_compact.c [hunk 3]
```diff
@@ -684,15 +684,22 @@ __wt_background_compact_signal(WT_SESSION_IMPL *session, const char *config)
     running = conn->background_compact.running;
 
     WT_ERR(__wt_config_getones(session, config, "background", &cval));
-    if (cval.val == running)
-        /*
-         * This is an error as we are already in the same state and reconfiguration is not allowed.
-         */
+    toggle = cval.val;
```
Comment (marcbutler) [inline src/conn/conn_compact.c:687]: Toggle appears to mean "enable"?

File: src/conn/conn_compact.c [hunk 4]
```diff
@@ -684,15 +684,22 @@ __wt_background_compact_signal(WT_SESSION_IMPL *session, const char *config)
...
+          session, EINVAL, "Cannot reconfigure background compaction while it's already running.");
+
+    /* If we haven't changed states, we're done. */
+    if (toggle == running)
+        goto err;
 
     /* Update the excluded tables when the server is turned on. */
-    if (cval.val) {
+    if (toggle) {
         __background_compact_exclude_list_clear(session, false);
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply src/conn/conn_compact.c:703]: That's a good point. We should only be able to update the `exclude` list when the background compaction is requested to start. If this stage fails, bg compaction should not start.

File: src/conn/conn_compact.c (hunk 3 above)
Comment (Sean04) [reply src/conn/conn_compact.c:687]: Yeah, is enable better?

**Comment by marcbutler** [comment]
Yeah I think `enable` would be clearer.

File: src/conn/conn_compact.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (Sean04) [reply src/conn/conn_compact.c:703]: We only reach this point if we're enabling background compaction from a disabled state. At this point, if we fail when parsing the exclude list below, we skip enabling the background compact server so it won't run.

[the remark this answers is not in the document]
Comment (marcbutler) [reply src/conn/conn_compact.c:703]: Yup got it.

**Review by marcbutler — APPROVED** [review]

**Comment by Sean04** [comment]
evergreen merge

**Pull request merged**